### ISSUE TITLE
How library linkings work?

### ISSUE THREAD

**Issue by milktea736** [nodejs/node-gyp#892]
Hi there, 

I have a wrap.node and it depends on two dynamic libraries, "testA.dylib" and "testB.dylib".
I can set the linking in binding.gyp. Once the wrap.node is generated, how the wrap.node interacts with these two libraries? I have this question due to the description, ["Node.js will first attempt to locate and load modules or JavaScript files that happen to share the same base name."](https://nodejs.org/api/addons.html#addons_addons). The wrap.node seems like a interface that connect javascript and binary code(two C++ libraries) to me. Am I get it right? 

Below are two different tests, I did set corresponding link paths in binding.gyp.
test.js requires the wrap.node, and call the methods in dynamic libraries.
Why is the first setting throws this:
module.js:440
  return process.dlopen(module, path._makeLong(filename));
Error: dlopen(/Users/test/FirstFolder/SecondFolde/wrap.node, 1): Library not loaded: testA.dylib

+FirstFolder (Doesn't work)
+=====   SecondFolder
+==========  wrap.node
+==========  test.js
+==========  dylibFolder
+===============  testA.dylib
+===============  testB.dylib

+FirstFolder (Works fine)
+=====   SecondFolder
+========== wrap.node
+========== test.js
+========== dylibFolder
+========== testA.dylib
+========== testB.dylib

Can I do any setting that make the first one works? 
I'm not sure whether this is relative to node-gyp or some basic rules how node.js works. 
Apologize if I post it in a wrong place. 
Appreciated for any helps.

**Comment by bnoordhuis** [comment]
For your wrap.node, it works the same way as any other shared object (which a .node file is, it's just a .dylib with a different extension), you link it to the libraries it depends on through the 'libraries' section in your binding.gyp:
```python
{
  'targets': [
    {
      'target_name': 'wrap',
      # ...
      'link_settings': {
        'libraries': ['-L/path/to/libraries', '-lA', '-lB'],
      },
      # ...
    }
  ]
}
```
If the libraries are in non-default directories, add `-Wl,-rpath,/path/to/libraries` and note that it can take relative-to-your-library paths; google 'rpath origin' for examples.

**Comment by milktea736** [comment]
Hi @bnoordhuis,

Thanks for your reply. I finally figure my problem out. My wrap.node is used in electron. The situation is, wrap.node works fine when I executing in terminal. After I built it as an application, my wrap.node can't link to the dynamic libraries when I click the application icon. The default dynamic library path of wrap.node is something like` "testA.dylib"`. I change it with command-line tool `"install_name_tool -change"` into a relative path. When I click the application icon, the runtime path is somehow out of my application bundle instead of staying with electron executable file. I added ` "process.chdir(__dirname);" ` in my main.js to back to my working directory. Finally, it links to dynamic library perfectly. Apologize for I get my problem completely wrong. Appreciate for your reply again.

**Comment by bnoordhuis** [comment]
No problem, happy to help.  I'll close the issue.

**Comment by milktea736** [comment]
Hi @mhomam, 
refers to you post #1404.
I assumed your env is something like this:
```
main.js
node_modules
  + mylib
    ++ mylib.node
    ++ sample.dylib
```
The reason is that `mylib` can't locate sample.dylib at runtime. If you require your module in `main.js`, 
node expects `sample.dylib` occuring in the same folder. 
```
# This might work for you without changing codes.
main.js
sample.dylib
node_modules
  + mylib
    ++ mylib.node
```

You should change the run time dependencies path to work in this structure:
```
main.js
node_modules
  + mylib
    ++ mylib.node
    ++ sample.dylib
```

type` otool -L mylib.node` will show you the library dependencies and run time paths. 
```
# demo depends on my_lib_A and my_lib_B
otool -L demo.node 
demo.node:
	my_lib_A.0.dylib (compatibility version 0.10.0, current version 0.10.0)
	my_lib_B.0.dylib (compatibility version 0.9.0, current version 0.9.3)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```

In my situation:
change the run time path of my_lib_A.0.dylib and my_lib_B.0.dylib
to 
```
demo.node:
	node_modules/demo/my_lib_A.0.dylib (compatibility version 0.10.0, current version 0.10.0)
	node_modules/demo/my_lib_B.0.dylib (compatibility version 0.9.0, current version 0.9.3)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```
with `install_name_tool -change "my_lib_A.0.dylib " "node_modules/demo/my_lib_A.0.dylib " demo.node` helps your `mylib` locating sample.dylib. Noted, if my_lib_A depends on my_lib_B, you should do the same to my_lib_A.dylib.
BTW, If you packing this on Windows, put all your libraries with `mylib.node` then everything works fine.

**Comment by tonetechnician** [comment]
Hey there,

II'm wondering if anyone on this thread has run into a similar linking problem on Windows? I've been trying to link a .dll for a good 2 days now and I'm still unsure exactly how to do this. There doesn't seem to be any clear cut method that I can find. 

So far, I've tried putting the .dll into the directory where my index.js is which doesn't work. 

I've tried linking my .dll into the libraries field, except it just throws the error that the file can't be read. 

I've also tried use he rpath syntax but it just doesn't seem to parse properly. 

```.sh
LINK : warning LNK4044: unrecognized option '/L\Windows\System32\.lib'; ignored [D:\git-repos\SAF-nodeaddon\build\SAF_native.vcxproj]
LINK : fatal error LNK1104: cannot open file 'saf_mkl_custom.lib' [D:\git-repos\SAF-nodeaddon\build\SAF_native.vcxproj]
```

Any help would be greatly appreciated, am feeling pretty stuck at this point.
